pkg/local_object_storage/engine/gc_test.go

```diff
@@ -165,3 +167,114 @@ func checkObjectsAsyncRemoval(t *testing.T, e *StorageEngine, cnr cid.ID, objs .
+			shard.WithExpiredLocksCallback(e.processExpiredLocks),
+			shard.WithGCRemoverSleepInterval(100*time.Millisecond),
+		)
+		require.NoError(t, err)
+	}
+	require.NoError(t, e.Open())
+	require.NoError(t, e.Init())
+	t.Cleanup(func() { _ = e.Close() })
+
+	cnr := cidtest.ID()
+
+	t.Run("expired object", func(t *testing.T) {
+		obj := generateObjectWithCID(cnr)
+		expAttr := objectSDK.NewAttribute(objectSDK.AttributeExpirationEpoch, fmt.Sprint(currEpoch))
+		obj.SetAttributes(expAttr)
+		require.NoError(t, e.Put(obj, nil))
+
+		e.HandleNewEpoch(currEpoch + 1)
+
+		checkObjectsAsyncRemoval(t, e, cnr, obj.GetID())
+	})
+
+	t.Run("object with lock - prevents deletion until lock expires", func(t *testing.T) {
+		const lockExpires = currEpoch + 2
+		e.HandleNewEpoch(currEpoch) // reset epoch to currEpoch
+
+		obj := generateObjectWithCID(cnr)
+		obj.SetAttributes(objectSDK.NewAttribute(objectSDK.AttributeExpirationEpoch, fmt.Sprint(currEpoch)))
+		require.NoError(t, e.Put(obj, nil))
+		objAddr := object.AddressOf(obj)
+
+		lockObj := generateObjectWithCID(cnr)
+		lockObj.SetType(objectSDK.TypeLock)
+		lockObj.SetAttributes(objectSDK.NewAttribute(objectSDK.AttributeExpirationEpoch, fmt.Sprint(lockExpires)))
+		require.NoError(t, e.Put(lockObj, nil))
+
+		require.NoError(t, e.Lock(cnr, lockObj.GetID(), []oid.ID{objAddr.Object()}))
+
+		es.e = currEpoch + 1
+		e.HandleNewEpoch(currEpoch + 1)
+
+		// object must stay because it's locked
+		_, err := e.Get(objAddr)
+		require.NoError(t, err)
+
+		es.e = lockExpires + 1
+		e.HandleNewEpoch(lockExpires + 1)
+
+		// now GC must delete the object eventually
+		require.Eventually(t, func() bool {
+			_, err := e.Get(objAddr)
+			return errors.Is(err, statusSDK.ErrObjectNotFound)
+		}, 2*time.Second, 200*time.Millisecond)
+	})
+
+	t.Run("expired tombstone removed", func(t *testing.T) {
+		tomb := generateObjectWithCID(cnr)
+		tomb.SetType(objectSDK.TypeTombstone)
+		tomb.SetAttributes(objectSDK.NewAttribute(objectSDK.AttributeExpirationEpoch, fmt.Sprint(currEpoch)))
+		require.NoError(t, e.Put(tomb, nil))
+
+		tombAddr := object.AddressOf(tomb)
+
+		e.HandleNewEpoch(currEpoch + 1)
+
+		require.Eventually(t, func() bool {
+			_, err := e.Get(tombAddr)
+			return errors.Is(err, statusSDK.ErrObjectNotFound)
+		}, 2*time.Second, 100*time.Millisecond)
+	})
+
+	t.Run("object covered by tombstone", func(t *testing.T) {
+		obj := generateObjectWithCID(cnr)
+		require.NoError(t, e.Put(obj, nil))
+		tomb := generateObjectWithCID(cnr)
+		tomb.SetType(objectSDK.TypeTombstone)
+
+		objAddr := object.AddressOf(obj)
+		tombAddr := object.AddressOf(tomb)
+
+		require.NoError(t, e.Inhume(tombAddr, 1, objAddr))
+
+		_, err := e.Get(objAddr)
+		require.Error(t, err, statusSDK.ErrObjectAlreadyRemoved)
+
+		e.HandleNewEpoch(currEpoch + 1)
+
+		require.Eventually(t, func() bool {
+			_, err := e.Get(objAddr)
+			return errors.Is(err, statusSDK.ErrObjectNotFound)
+		}, 3*time.Second, 100*time.Millisecond)
```

Comment:
why wait time and interval varies? We have `1s` (`checkObjectsAsyncRemoval`), `2s` and `3s`; `100ms` and `200ms`